Code generator: use different arrays for constants, computed constants, and algebraic variables

agarny opened the pull request:
Fixes #1243.

agarny committed:
Code generator: use different arrays for constants, computed constants, and algebraic variables.

agarny committed:
Python docstrings tests: don't test for CellmlElementType.

agarny committed:
Generator: use different arrays for constants, computed constants, and algebraic variables for initialiseVariables().

agarny committed:
Generator: use different arrays for constants, computed constants, and algebraic variables for computeRates().

agarny committed:
Generator: use different arrays for constants, computed constants, and algebraic variables for computeVariables().

agarny committed:
Generator: use different arrays for constants, computed constants, and algebraic variables for computeComputedConstants().

agarny committed:
Generator: use computed_constants and not computedConstants in Python.

agarny committed:
Generator profile: bumped the C and Python versions.

agarny committed:
Generator profile: added setters/getters for constants, computed constants, and algebraic arrays.

agarny committed:
Tests: automatically generate the expected file contents.

To do it manually is very time consuming while here it gets done automatically. From there, we can quickly confirm, using git, whether the new expected file contents is correct.

agarny committed:
Analyser: track the constant, computed constant, and algebraic indexes.

agarny committed:
Some minor cleaning up.

agarny committed:
Generator: added an external array.

agarny committed:
Updated JavaScript tests.

agarny committed:
Analyser: keep track of computed constants and external variables for an equation.

We will also need to keep track of states and constants.

agarny committed:
Analyser: some minor cleaning up.

agarny committed:
AnalyserModel: keep track of the constants, computed constants, and external variables.

agarny committed:
AnalyserEquation: keep track of states and constants.

agarny committed:
Generator: generate the CONSTANT_INFO, COMPUTED_CONSTANT_INFO, ALGEBRAIC_INFO, and EXTERNAL_INFO arrays rather than just the VARIABLE_INFO array.

agarny committed:
Generator: don't need to mention the type of a variable in a XXX_INFO array anymore.

agarny committed:
GeneratorProfile: extracted generator profile_p.h.

agarny committed:
Generator: updated the createXxxArray() methods.

agarny committed:
Generator: updated the signature of our initialiseVariables() method.

agarny committed:
GeneratorProfile: fixed a small issue with our Python profile.

agarny committed:
Generator: initialise things in initialiseVariables() in the order of its parameters.

agarny committed:
Generator: initialise all computed constants in computeComputedConstants().

agarny committed:
GeneratorProfile: added the external array string.

agarny committed:
Generator: don't initialise our external variables in initialiseVariables().

There was never a need to initialise them there in the first place. In fact, an external variable is effectively an algebraic variable which is computed by an end user. So, it needs to be computed in computeRates() (if available) and in computeVariables().

Also fixed an issue with initialiseVariables() not always including the VOI.

agarny committed:
Analyser: a variable using an NLA should always be an algebraic variable.

Even if the NLA equation only uses constant variables and constant values.

agarny committed:
Analyser: some minor cleaning up.

Constants are indexed using makeConstant().

agarny committed:
Analyser: properly set the index of external variables.

agarny committed:
Analyser: put some struct/class definitions in a corresponding private header.

agarny committed:
Analyser: simplified things.

agarny committed:
Generator: make sure that constants are initialised when they are needed to initialise a state.

agarny committed:
Generator: renamed the "external" array to "externals".

Oversight on my part!

agarny committed:
Generator: "initialise" computed constants in initialiseVariables().

This effectively reverts commit b1b75362.

Indeed, some computed constants are initialised using an equation (e.g., x = 3 rather than x with an initial value of 3). So, there is no need to compute those computed constants many times (using computeComputedConstants()). Instead, it's much better to compute them in initialiseVariables().

Not to mention that we might have "computedConstants[1] = 1.0;" and then "computedConstants[0] = computedConstants[1];" so if we were to compute both of them in computeComputedConstants(), we would need to compute "computedConstants[1] = 1.0;" before computing "computedConstants[0] = computedConstants[1];", which would make things slightly more complicated to generate.

agarny committed:
Generator: initialise things in initialiseVariables() in the order of its parameters.

agarny committed:
GeneratorProfile: pass the computedConstants array to initialiseVariables().

agarny committed:
GeneratorProfile: don't need to pass the VOI to initialiseVariables().

agarny committed:
Tests: removed unneeded generated C/Python files.

agarny committed:
Generator: make sure that constants are initialised when they are needed to initialise a state.

agarny committed:
GeneratorProfile: use the constants, computed constants, and algebraic arrays rather than the variables array with findRoot() & Co.

agarny committed:
GeneratorProfile: use the constants, computed constants, and algebraic arrays rather than the variables array to compute external variables.

agarny committed:
GeneratorProfile: use the externals array to retrieve the value of external variables.

agarny committed:
Tests: some minor cleaning up.

agarny committed:
Addressed various coverage issues as a result of our work on issue #1247.

agarny committed:
Some minor cleaning up.

agarny committed:
Analyser: improved unit scaling.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Generator: make sure that NLA systems have access to external variables.

agarny committed:
Generator: fixed a small issue with our Python profile.

agarny committed:
Tests: added a test to generate some code for the DAE version of the HH52 model with various external variables.

agarny committed:
Tests: changed the external variable in algebraicEqnWithOneNonIsolatedUnknownWithExternalVariable.

This is so that we can test an algebraic model with an external variable (in https://github.com/agarny/libCellMLGeneratedCodeTests).

agarny committed:
Python: improved the API documentation.

agarny committed:
Some minor cleaning up.

agarny committed:
Merge branch 'main' into issue1243.

agarny committed:
Merge branch 'issue1251' into issue1243

agarny committed:
Some minor cleaning up.

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'issue1251' into issue1243

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Some minor cleaning up.

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'issue1251' into issue1243

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'main' into issue1243

agarny committed:
Merge branch 'issue1251' into issue1243

agarny committed:
Some minor cleaning up following the last merge.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Fixed macro issue on some Linux systems.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Merge branch 'issue1251' into issue1243.

agarny committed:
Reverted commit 6a48c0fa8.

agarny committed:
Merge branch 'main' into issue1243.

agarny committed:
Merge branch 'main' into issue1243.

agarny committed:
Merge branch 'main' into issue1243.

agarny committed:
Merge branch 'main' into issue1243

hsorby commented on src/bindings/interface/analyserequation.i:
```diff
@@ -43,14 +43,41 @@
+"Returns the algebraic variables computed by this :class:`AnalyserEquation` object.";
+
+%feature("docstring") libcellml::AnalyserEquation::algebraic
+"Returns the algebraic variable, at the given index, computed by this :class:`AnalyserEquation` object.";
```
This looks like it is referencing the same target, how does it differentiate the overloads?

agarny replied:
Yes, I am not quite sure indeed. Anyway, it should be fine once I have renamed `algebraic` to `algebraicVariables` in which case we will then have two bindings: `algebraicVariable(index)` and `algebraicVariables()`.

agarny replied:
Fixed in f76d0b19d.

agarny committed:
Merge branch 'main' into issue1243.

agarny committed:
Merge branch 'issue1333' into issue1243

hsorby commented on tests/test_utils.cpp:
```diff
@@ -417,3 +416,19 @@ void compareReset(const libcellml::ResetPtr &r1, const libcellml::ResetPtr &r2)
+void expectEqualFileContents(const std::string &fileName, const std::string &fileContents)
+{
+    // Uncomment the below when you want to generate the expected file contents.
+    // #define NEW_GENERATOR
```
This sort of thing should be done with a CMake command so you can run a target that updates all the file contents.

agarny replied:
Maybe it should be done, but I went for the most convenient option (for me, at least). I mean, it's like the CMake target you created to update the SHA-1 values. I have never used it because... I don't find it convenient (not least because [we still don't list available targets upon configuration](https://github.com/cellml/libcellml/issues/1173), so I had to keep looking for the target name; working on different CMake-based projects with different CMake targets, we can't expect people to remember everything). Here, all I have to do is to uncomment one line of code and I am done.

hsorby replied:
Sure, I do like convenient. Now all you have to do is remember that there is a random line of code that you can uncomment to do what you want, somewhere amongst hundreds of thousands of lines of code that you work on. You cannot interrogate a limited list of targets in the build directory and maybe get a clue for what you are looking for.

agarny replied:
I have my notes that help me for that, so I don't need to remember anything (same for the SHA-1 values), but I accept that it's not ideal either, just more convenient and faster for me (especially when I need to update generated files a lot, like right now).

Anyway, feel free to create that CMake target if you have nothing better to do.

hsorby commented on src/api/libcellml/analyserequation.h:
```diff
@@ -166,33 +166,120 @@ class LIBCELLML_EXPORT AnalyserEquation
+     *
+     * @return The algebraic variables as a @c std::vector.
+     */
+    std::vector<AnalyserVariablePtr> algebraic() const;
```
I wouldn’t mind this API being `algebraicVariables` I have seen it in the javascript bindings and I prefer it to this.  The same applies to `algebriacCount` and `algebraic` for getting an algegraic variable.

agarny replied:
Fixed in f76d0b19d.

agarny committed:
Merge branch 'main' into issue1243.

hsorby commented on src/generator.cpp:
```diff
@@ -76,50 +76,52 @@ AnalyserVariablePtr Generator::GeneratorImpl::analyserVariable(const VariablePtr
-                }
-            } while (res == nullptr);
-        }
+        } while (res == nullptr);
```
This should just be:

```
auto modelVariables = variables(mModel);
res = *std::find_if(modelVariables.begin(), modelVariables.end(), [=](const auto &modelVariable) {
            return mModel->areEquivalentVariables(modelVariable->variable(), variable);
        });
```

agarny replied:
Fixed in 01ab5935d.

hsorby commented on src/generator.cpp:
```diff
@@ -76,50 +76,52 @@ AnalyserVariablePtr Generator::GeneratorImpl::analyserVariable(const VariablePtr
+        auto initialValueAnalyserVariable = Generator::GeneratorImpl::analyserVariable(initialValueVariable);
+
+        if (owningComponent(variable) != owningComponent(initialValueAnalyserVariable->variable())) {
+            return Units::scalingFactor(initialValueVariable->units(), variable->units());
```
I’m not a fan of calling the same function more than once, I preger to assign a variable:

```
        auto variableComponent = owningComponent(variable);
        auto initialValueVariable = variableComponent->variable(variable->initialValue());
        auto initialValueAnalyserVariable = Generator::GeneratorImpl::analyserVariable(initialValueVariable);
        if (variableComponent != owningComponent(initialValueAnalyserVariable->variable())) {
            return Units::scalingFactor(initialValueVariable->units(), variable->units());
        }

```

agarny replied:
Fixed in 94c54689f.

hsorby commented on src/generator.cpp:
```diff
@@ -398,147 +405,98 @@ void Generator::GeneratorImpl::addVariableInfoObjectCode()
 
-void Generator::GeneratorImpl::addImplementationStateInfoCode()
+void Generator::GeneratorImpl::doAddImplementationVariableInfoCode(const std::string &variableInfoString,
+                                                                   const std::vector<AnalyserVariablePtr> &variables,
```
I think this needs to be : analyserVariables
plus the adjustment for the flow on effects.

agarny replied:
Fixed in 0114a6137.

hsorby commented on src/generator.cpp:
```diff
@@ -398,147 +405,98 @@ void Generator::GeneratorImpl::addVariableInfoObjectCode()
-        }
+    doAddImplementationVariableInfoCode(mProfile->implementationConstantInfoString(), mModel->constants(), false);
+    doAddImplementationVariableInfoCode(mProfile->implementationComputedConstantInfoString(), mModel->computedConstants(), false);
+    doAddImplementationVariableInfoCode(mProfile->implementationAlgebraicInfoString(), mModel->algebraic(), false);
```
Is there a reason why we don’t add this code conditionally? Like we do for external variables?

agarny replied:
Yes, there is: a model doesn't always have a VOI / states (algebraic/NLA model vs. an ODE/DAE model) and/or external variables (something that an end user decides to have, or not), hence we check whether the model has a VOI / states and/or external variables and generate code for them, or not.

When it comes to constants, computed constants, and algebraic variables, it's not like the type of a model will tell us whether a model has constants, computed constants, and/or algebraic variables. It's very much dependent on a model itself, hence we always generate information for any potential constants, computed constants, and algebraic variables.

hsorby replied:
I thought so, I was thinking that we know if the code is empty therefore unnecessary. I was wondering more from the consequences of people using the generated code. Does it make life more difficult or unreasonably so if functions/information strings are sometimes present and sometimes not, rather than empty.

agarny replied:
Yes, I have been wondering the same, but... remember that the signature of some functions depends on whether the model is a differential model and/or whether it has external variables (e.g., see [here](https://github.com/agarny/libcellml/blob/issue1243/src/generatorprofile.cpp#L2900-L2935)). And, when it comes to the signature of those functions, I would find it odd if the signature was to include `voi`, `states`, `rates`, `external_variables`, and `external_variable` if they were to be of no use.

Still, happy to discuss this further at a 4:15pm meeting. :)

hsorby commented on src/generator.cpp:
```diff
@@ -398,147 +405,98 @@ void Generator::GeneratorImpl::addVariableInfoObjectCode()
-                break;
-            }
-
-            auto variableVariable = variable->variable();
+    if (modelHasOdes()) {
+        doAddImplementationVariableInfoCode(mProfile->implementationVoiInfoString(), {mModel->voi()}, true);
+    }
```
Why the two `if (modelHasOdes()) {`? Can they not be combined into one?

agarny replied:
Agreed, fixed in 7b73df6a0.

hsorby commented on src/generator.cpp:
```diff
@@ -1757,9 +1752,9 @@ std::string Generator::GeneratorImpl::generateEquationCode(const AnalyserEquatio
+            if (!mProfile->findRootCallString(modelHasOdes(), mModel->hasExternalVariables()).empty()) {
                 res += mProfile->indentString()
-                       + replace(mProfile->findRootCallString(modelHasOdes()),
+                       + replace(mProfile->findRootCallString(modelHasOdes(), mModel->hasExternalVariables()),
```
Why call modelHasOdes multiple times?

agarny replied:
Fixed in c8ffaf27c.

hsorby commented on src/generator_p.h:
```diff
@@ -80,30 +80,27 @@ struct Generator::GeneratorImpl
-    void addInterfaceVoiStateAndVariableInfoCode();
-    void addImplementationVoiInfoCode();
-    void addImplementationStateInfoCode();
+    void doAddImplementationVariableInfoCode(const std::string &variableInfoString,
```
Using ‘do’ as a prefix has been used in this codebase to indicate that this is the implementation via a virtual function. Keeping virtual methods out of the API is better for binary stability. This function is not virtual and I think should make use of C++ override functionality instead.

agarny replied:
Agreed, C++ overloading functionality could be used here. Fixed in 8c5ef05a8.

hsorby commented on src/utilities.h:
```diff
@@ -870,4 +870,26 @@ size_t mathmlChildCount(const XmlNodePtr &node);
+ *
+ * @return The variables in the given model.
+ */
+std::vector<AnalyserVariablePtr> variables(const AnalyserModelPtr &model);
```
I really think this function (and the next) should be called `analyserVariables`. This function retuns a vector of analyser variables which in libCellML are a different concept from variables.

agarny replied:
Agreed, fixed in bbc20a9d8.

hsorby reviewed (commented):
I not sure about the API for the getting the variable types. For me it should be algebraicVariables and not algebraic. Similar for the related API. I would be interested in what Andre thinks of this. 

How do people feel about the API consistency we have create_algebriac_array() but initialise_variables(), which initialises states and constants. Do we just need initialise_arrays().

I would like to have a discussion around the API and use of `variables` for both componet variables and analyser variables. I myself would prefer to have the two distinct. For me it muddies the clarity of the code when a variable might be a different object depending on where it has come from.

agarny commented:
> I not sure about the API for the getting the variable types. For me it should be algebraicVariables and not algebraic. Similar for the related API. I would be interested in what Andre thinks of this.

I  can't argue about preferring `algebraicVariables` to `algebraic` since this has been bugging me for quite some time. If I recall correctly, the CellML API uses `algebraic` so I originally went with that, but... yeah!

> How do people feel about the API consistency we have create_algebriac_array() but initialise_variables(), which initialises states and constants. Do we just need initialise_arrays().

Not keen on `initialise_arrays()`, but agree that `initialise_variables()` is not great. What about `initialise_model()`?

> I would like to have a discussion around the API and use of `variables` for both componet variables and analyser variables. I myself would prefer to have the two distinct. For me it muddies the clarity of the code when a variable might be a different object depending on where it has come from.

Can discuss that on Tuesday, if you want. (I am offline tomorrow).

agarny committed:
Renamed algebraic and external variables.

agarny committed:
Generator: improved the implementation of Generator::GeneratorImpl::analyserVariable().

agarny committed:
Generator: use temporary variable.

agarny committed:
Generator: use a more suitable parameter name in Generator::GeneratorImpl::doAddImplementationVariableInfoCode().

agarny committed:
Generator: simplified the implementation of Generator::GeneratorImpl::addImplementationVariableInfoCode().

agarny committed:
Generator: slight improvement to Generator::GeneratorImpl::generateEquationCode().

agarny committed:
Generator: use C++ overloading functionality.

agarny committed:
Utilities: renamed our `variables()` utility to `analyserVariables()`.

agarny commented:
@hsorby / @nickerso: I believe I have addressed Hugh's issues.

hsorby commented on src/generator.cpp:
```diff
@@ -398,147 +388,95 @@ void Generator::GeneratorImpl::addVariableInfoObjectCode()
             }
 
-            auto stateVariable = state->variable();
+            auto variableVariable = analyserVariable->variable();
```
Not fussed on this one but should the variable name here be changed to `analyserVariableVariable`?

agarny replied:
Agreed, fixed in e46d26745.

hsorby commented:
> Can discuss that on Tuesday, if you want. (I am offline tomorrow).

I feel this may be a 4.15 Friday discussion?

agarny committed:
Generator: use `analyserVariable` rather than `variable` to distinguish between our two types of variables.

agarny committed:
Generator: replaced `initialiseVariables()` with `initialiseArrays()`.

This is more consistent with our `createXxxArray()` methods and `deleteArray()` method.

agarny commented:
> > How do people feel about the API consistency we have create_algebriac_array() but initialise_variables(), which initialises states and constants. Do we just need initialise_arrays().
> Not keen on `initialise_arrays()`, but agree that `initialise_variables()` is not great. What about `initialise_model()`?

Just remembered that we haven't got a consensus on this one. Thinking more about it, I guess we could have `initialiseArrays()`. At least, it would be consistent with our `createXxxArray()` methods and our `deleteArray()` method. Also, I think it would make more sense to use `initialiseComputedConstants

Ok, done in b4045c6cd.

agarny commented:
By the way, we replaced instances like `const AnalyserVariablePtr &variable` with `const AnalyserVariablePtr &analyserVariable`, but if anything this has made our code less consistent. Indeed, if we are to do that then we should do something similar with things like `const AnalyserModelPtr &model`, `const AnalyserEquationPtr &equation`, `const AnalyserEquationAstPtr &equationAst`, `const AnalyserExternalVariablePtr &externalVariable`, etc. and... also with the internal variants (e.g., `const AnalyserInternalVariablePtr &variable`).

So... I guess I should replace all of those too...? @hsorby?

hsorby commented:
> By the way, we replaced instances like `const AnalyserVariablePtr &variable` with `const AnalyserVariablePtr &analyserVariable`, but if anything this has made our code less consistent. Indeed, if we are to do that then we should do something similar with things like `const AnalyserModelPtr &model`, `const AnalyserEquationPtr &equation`, `const AnalyserEquationAstPtr &equationAst`, `const AnalyserExternalVariablePtr &externalVariable`, etc. and... also with the internal variants (e.g., `const AnalyserInternalVariablePtr &variable`).
> 
> So... I guess I should replace all of those too...? @hsorby?

I think this PR is heavy enough as is and maybe the case for a new issue and then do this separately to this PR.

agarny commented:
> I think this PR is heavy enough as is and maybe the case for a new issue and then do this separately to this PR.

Fair enough. Would then be good to merge this PR ASAP, so that I can submit a follow-up PR.

hsorby commented on src/analyser_p.h:
```diff
@@ -0,0 +1,264 @@
+limitations under the License.
+*/
+
+// #include "libcellml/analyser.h"
```
We can remove this since it isn’t required.

agarny replied:
Oops, fixed in cc143c442.

hsorby commented on src/generator.cpp:
```diff
@@ -930,8 +899,14 @@ std::string Generator::GeneratorImpl::generateVariableNameCode(const VariablePtr
         arrayName = state ?
                         mProfile->statesArrayString() :
                         mProfile->ratesArrayString();
+    } else if (analyserVariable->type() == AnalyserVariable::Type::CONSTANT) {
```
Don’t you just want to call this once: `analyserVariable->type()`? And not five times? I know I love a good assigned variable but it makes sense here right.

agarny replied:
I find this pointless, but done in ab1866385 since you have already asked me to do something similar before.

Still, I thought I would ask Claude Sonnet 4.5 what it thinks:

agarny: Is it better to have:
```
std::string Generator::GeneratorImpl::generateVariableNameCode(const VariablePtr &variable,
                                                               bool state) const
{
    // Generate some code for a variable name, but only if we have a model. If
    // we don't have a model, it means that we are using the generator from the
    // analyser, in which case we just want to return the original name of the
    // variable.

    if (mModel == nullptr) {
        return variable->name();
    }

    auto analyserVariable = Generator::GeneratorImpl::analyserVariable(variable);
    auto analyserVariableType = analyserVariable->type();

    if (analyserVariableType == AnalyserVariable::Type::VARIABLE_OF_INTEGRATION) {
        return mProfile->voiString();
    }

    std::string arrayName;

    if (analyserVariableType == AnalyserVariable::Type::STATE) {
        arrayName = state ?
                        mProfile->statesArrayString() :
                        mProfile->ratesArrayString();
    } else if (analyserVariableType == AnalyserVariable::Type::CONSTANT) {
        arrayName = mProfile->constantsArrayString();
    } else if (analyserVariableType == AnalyserVariable::Type::COMPUTED_CONSTANT) {
        arrayName = mProfile->computedConstantsArrayString();
    } else if (analyserVariableType == AnalyserVariable::Type::ALGEBRAIC_VARIABLE) {
        arrayName = mProfile->algebraicVariablesArrayString();
    } else {
        arrayName = mProfile->externalVariablesArrayString();
    }

    return arrayName + mProfile->openArrayString() + convertToString(analyserVariable->index()) + mProfile->closeArrayString();
}
```
or have:
```
std::string Generator::GeneratorImpl::generateVariableNameCode(const VariablePtr &variable,
                                                               bool state) const
{
    // Generate some code for a variable name, but only if we have a model. If
    // we don't have a model, it means that we are using the generator from the
    // analyser, in which case we just want to return the original name of the
    // variable.

    if (mModel == nullptr) {
        return variable->name();
    }

    auto analyserVariable = Generator::GeneratorImpl::analyserVariable(variable);

    if (analyserVariable->type() == AnalyserVariable::Type::VARIABLE_OF_INTEGRATION) {
        return mProfile->voiString();
    }

    std::string arrayName;

    if (analyserVariable->type() == AnalyserVariable::Type::STATE) {
        arrayName = state ?
                        mProfile->statesArrayString() :
                        mProfile->ratesArrayString();
    } else if (analyserVariable->type() == AnalyserVariable::Type::CONSTANT) {
        arrayName = mProfile->constantsArrayString();
    } else if (analyserVariable->type() == AnalyserVariable::Type::COMPUTED_CONSTANT) {
        arrayName = mProfile->computedConstantsArrayString();
    } else if (analyserVariable->type() == AnalyserVariable::Type::ALGEBRAIC_VARIABLE) {
        arrayName = mProfile->algebraicVariablesArrayString();
    } else {
        arrayName = mProfile->externalVariablesArrayString();
    }

    return arrayName + mProfile->openArrayString() + convertToString(analyserVariable->index()) + mProfile->closeArrayString();
}
```
I mean, in the first case, we create a temporary variable while in the second case I would expect modern compilers to be clever enough to inline the call to `analyserVariable->type()` since all it does is return `mPimpl->mType`?

GitHub Copilot: You're absolutely right to think about this! For modern compilers, the **second version** (without the temporary `analyserVariableType` variable) is generally **equally good or slightly better**. Here's why:

## Performance Perspective

1. **Method call overhead**: If `analyserVariable->type()` is a simple getter that returns `mPimpl->mType`, modern compilers (especially with optimizations enabled like `-O2` or `-O3`) will almost certainly inline this call. This means there's essentially zero overhead.

2. **Caching consideration**: In the first version, you're storing the result to avoid calling `type()` multiple times. However, since it's just a member access, the compiler's optimizer can recognize that `analyserVariable->type()` returns the same value each time (no side effects) and can cache it in a register automatically if needed.

3. **Code generation**: With optimization, both versions will likely generate identical or nearly identical machine code.

## Code Quality Perspective

The **second version is better** because:

1. **Less clutter**: One fewer variable to track mentally
2. **Clearer intent**: The code directly expresses "check the type" rather than "store the type, then check it"
3. **Smaller scope**: No intermediate variable that lives longer than needed
4. **More maintainable**: If the logic changes, there's no risk of forgetting to update the cached variable

## When the First Approach Would Matter

The temporary variable would be beneficial if:
- The method call was expensive (complex computation, I/O, locks, etc.)
- The method had side effects
- You're working without compiler optimizations (debug builds)
- The method call appears many times (say, 10+ times)

## Recommendation

**Go with the second version**. It's cleaner, more idiomatic C++, and compilers are very good at optimizing simple getters. The performance difference is negligible, and the readability improvement is worth it.

If you're curious about the actual generated code, you could always check the assembly output with optimization flags to confirm they're equivalent, but I'd be very surprised if they weren't.

hsorby replied:
Should have left it then, can’t go against AI. Although I would be tempted to check it’s claims about the same machine code, which for me is the major reason I thought creating a local variable was better than calling a function multiple times. That thinking may well be a hangover from compilers of the past and may need updating to be inline with current compilers.

agarny replied:
I know that I am not bothered about calling those small methods and making those methods `inline`. Compilers have been clever enough to inline them for us automatically for a long time now. But, granted, I have never looked into the generated machine code. It's just something that I have seen and read many times over the years.

Anyway, I could revert those changes... or leave them as-is. Views?

hsorby replied:
I think probably revert, but I will check that statement by the AI.

agarny replied:
Ok, reverted (and force pushed).

hsorby replied:
I generated the assembly code for both implementations and … they compiled to exactly the same assembly code.  So that just leaves readability as the deciding factor. And that is opinion based.

hsorby commented on src/generator.cpp:
```diff
@@ -1931,10 +1922,12 @@ void Generator::GeneratorImpl::addImplementationComputeRatesMethodCode(std::vect
             // NLA equation in case the rate is not on its own on either the LHS
             // or RHS of the equation.
 
+            auto variables = libcellml::analyserVariables(equation);
```
Why are these function calls given the libcellml namespace? Does some compiler need them?

agarny replied:
That's because it used to read `auto variables = libcellml::variables(equation);`, but the `variables()` method got renamed to `analyserVariables()`. I then did a "quick" a find&replace and clearly missed the name of the variable.

Anyway, fixed in f92caf5c4.

hsorby reviewed (commented):
analyser_p.h needs adding to GIT_HEADER_FILES

Some minor edits for me then I think it will ge all good.

agarny committed:
src/analyser_p.h: removed unneeded header.

agarny committed:
CMake: added `analyser_p.h` to `GIT_HEADER_FILES`.

agarny committed:
Analyser/Generator: some cleaning up around the `libcellml` namespace.

agarny commented:
> analyser_p.h needs adding to GIT_HEADER_FILES

Oops, done in 37e2b7b34.

agarny commented:
Ok, I believe we are good to go @hsorby and @nickerso.
